### PR TITLE
AAU, on the CLI, I see a new method for importing assets

### PR DESCRIPTION
**Pull request by theodu**
- [ ] I opened a merge request on `kili` on a branch with the same name than the branch of the current pull request and forced security tests.
- [ ] I opened a pull request on the Node playground if the changes are breaking.

### CODE REVIEW

File: kili/cli.py [hunk 1]
```diff
@@ -12,8 +14,70 @@ def cli():
...
+@project.command(name='import')
+@click.argument('files', type=str, nargs=-1)
+@click.option('--api-key', type=str, help='Your Api Key', envvar='KILI_API_KEY')
+@click.option('--project-id', type=str, help='Id of the project to import assets in')
+@click.option('--exclude', '-e', type=str, multiple=True,
+              help="Files to exclude from the given files")
+@click.option('--frames', type=bool, help="Only for a frame project, import videos as frames")
+@click.option('--fps', type=int, help="Only for a frame project, import videos with this fps")
+#pylint: disable=too-many-arguments
+def import_assets(api_key, project_id, files, exclude, fps, frames):
```
Comment (crsegerie) [inline kili/cli.py:31]: Why don't you use type annotation?

File: kili/cli.py [hunk 2]
```diff
@@ -12,8 +14,70 @@ def cli():
...
+
+    Currently, this command does not support :
+        - The import of videos from local frames, rich text and time series assets.
+        - the import of assets with metadata or with a custom external_id
+    For such imports, please use the `append_many_to_dataset` method in the Kili SDK.
+    """
+    if files is None:
+        raise ValueError("No files or directory specified.")
+    kili = Kili(api_key=api_key)
+    input_type = kili.projects(project_id, disable_tqdm=True, fields=[
```
Comment (crsegerie) [inline kili/cli.py:43]: bad formatting, it's difficult to read

File: kili/cli.py [hunk 3]
```diff
@@ -12,8 +14,70 @@ def cli():
...
+            'Check that the paths exist and that the file types are compatible with the project')
+
+    external_ids = [path.split('/')[-1] for path in files_to_upload]
+
+    json_metadata_array = None
+    if input_type == 'FRAME':
+        json_metadata_array = [
+            {'processingParameters': {
+                'shouldKeepNativeFrameRate': fps is None,
+                'framesPlayedPerSecond': fps,
```
Comment (crsegerie) [inline kili/cli.py:67]: fps can be None here

File: kili/mutations/asset/helpers.py [hunk 4]
```diff
@@ -77,16 +79,20 @@ def process_content(input_type: str,
...
-        return [content if is_url(content) else (content
-            if (json_content_array is not None and json_content_array[i] is not None)
-            else (encode_base64(content) if check_file_mime_type(content, input_type) else None))
-            for i, content in enumerate(content_array)]
+        return [content if is_url(content)
+                else (content
+                      if (json_content_array is not None and json_content_array[i] is not None)
+                      else (encode_base64(content) if check_file_mime_type(content, input_type)
+                            else None))
+                for i, content in enumerate(content_array)]
```
Comment (crsegerie) [inline kili/mutations/asset/helpers.py:87]: In order to avoid those types of formatting modifications which do not change anything, I would recommend black formatter

File: kili/cli.py (hunk 1 above)
Comment (theodu) [reply kili/cli.py:31]: I thought that it inherited from the type of the decorators above but indeed it seems that it doesn't

File: kili/mutations/asset/helpers.py [hunk 5]
```diff
@@ -272,4 +281,39 @@ def process_update_properties_in_assets_parameters(properties) -> dict:
...
+
+def get_file_to_upload(files: List[str], input_type: str, exclude: List[str]):
+    """Get files to upload when giving a list of file or folder paths
+
+    Args:
+        files: a list path that can either be file paths, folder paths or unexisting paths
+        input_type: input type of the project to import data to.
+
+    Returns:
+        a list of existing files to upload, compatible with the project type.
```
Comment (crsegerie) [inline kili/mutations/asset/helpers.py:294]: add `-> List[str] `to fonction defeinition

File: kili/cli.py (hunk 3 above)
Comment (theodu) [reply kili/cli.py:67]: This is wanted. If the fps option is not given, it will be None, then fps will be None and shouldKeepNativeFrameRate will be true. the video will be imported with its native fps.
Also, in `append_many_to_dataset`, if fps is none, it will be set to the default value which is 25

File: kili/mutations/asset/helpers.py (hunk 4 above)
Comment (theodu) [reply kili/mutations/asset/helpers.py:87]: Yes indeed, he have a ticket for that that we haven't taken yet: https://linear.app/kili-technology/issue/ML-33/aakd-i-want-the-python-sdk-code-formatted-with-black

File: kili/mutations/asset/helpers.py [hunk 6]
```diff
@@ -272,4 +281,39 @@ def process_update_properties_in_assets_parameters(properties) -> dict:
...
+    files_path = [item for item in files if os.path.isfile(
+        item)]
+    folders_to_search_in = [item for item in files if os.path.isdir(item)]
+    for folder_path in folders_to_search_in:
+        # make sure that it ends with a backslash
+        folder_path = os.path.join(folder_path, '')
+        files_path.extend([item for item in glob.glob(folder_path + '*')
+                           if os.path.isfile(item)])
+    files_path_to_upload = list(
+        filter(lambda content: check_file_mime_type(content, input_type, False), files_path))
```
Comment (crsegerie) [inline kili/mutations/asset/helpers.py:305]: list comprehension is generally less verbose than list(filter(lambda ...

File: test/test_cli.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        open('test_tree/image2.jpg', 'w')
+        open('test_tree/texte1.txt', 'w')
+        open('test_tree/video1.mp4', 'w')
+        open('test_tree/video2.mp4', 'w')
+        os.mkdir('test_tree/leaf')
+        open('test_tree/leaf/image3.png', 'w')
+        open('test_tree/leaf/image4.jpg', 'w')
+        open('test_tree/leaf/texte2.txt', 'w')
+
+        for test_case in TEST_CASES:
```
Comment (crsegerie) [inline test/test_cli.py:118]: Great design pattern 👍

File: kili/mutations/asset/helpers.py [hunk 8]
```diff
@@ -272,4 +281,39 @@ def process_update_properties_in_assets_parameters(properties) -> dict:
...
+        a list of existing files to upload, compatible with the project type.
+    """
+    files_path = [item for item in files if os.path.isfile(
+        item)]
+    folders_to_search_in = [item for item in files if os.path.isdir(item)]
+    for folder_path in folders_to_search_in:
+        # make sure that it ends with a backslash
+        folder_path = os.path.join(folder_path, '')
+        files_path.extend([item for item in glob.glob(folder_path + '*')
+                           if os.path.isfile(item)])
```
Comment (crsegerie) [inline kili/mutations/asset/helpers.py:303]: I think here the for loop could be simplified by a regex in glob, but it's ok

File: kili/cli.py [hunk 9]
```diff
@@ -12,8 +15,86 @@ def cli():
...
+
+        - the import of videos from local frames, rich text and time series assets
+
+        - the import of assets with metadata or with a custom external_id
+
+    For such imports, please use the `append_many_to_dataset` method in the Kili SDK.
+    """
+    if files is None:
+        raise ValueError("No files or directory specified.")
+    kili = Kili(api_key=api_key)
```
Comment (mDuval1) [inline kili/cli.py:58]: how can on premise customers change the endpoint ?

File: kili/cli.py [hunk 10]
```diff
@@ -12,8 +15,86 @@ def cli():
...
+    input_type = kili.projects(project_id,
```
Comment (mDuval1) [inline kili/cli.py:59]: What if the project does not exist ? Throw a better error

File: kili/cli.py [hunk 11]
```diff
@@ -12,8 +15,86 @@ def cli():
...
+                               disable_tqdm=True,
+                               fields=['inputType'])[0]['inputType']
+
+    if input_type != 'FRAME' and (fps is not None or frames is True):
```
Comment (mDuval1) [inline kili/cli.py:63]: First treat edge cases, then treat the general case (you're doing the opposite here)

File: kili/cli.py [hunk 12]
```diff
@@ -12,8 +15,86 @@ def cli():
...
+
+    Files can be paths to files or to folders. You can provide several paths separated by spaces.
+
+    Currently, this command does not support:
+
+        - the import of videos from local frames, rich text and time series assets
+
+        - the import of assets with metadata or with a custom external_id
+
+    For such imports, please use the `append_many_to_dataset` method in the Kili SDK.
```
Comment (mDuval1) [inline kili/cli.py:54]: Add CLI examples

File: kili/mutations/asset/helpers.py [hunk 13]
```diff
@@ -272,4 +281,48 @@ def process_update_properties_in_assets_parameters(properties) -> dict:
     properties['should_reset_to_be_labeled_by_array'] = list(map(
         is_none_or_empty, properties['to_be_labeled_by_array']))
     return properties
-    
\ No newline at end of file
+
+
+def get_file_to_upload(files: Tuple[str, ...],
```
Comment (mDuval1) [inline kili/mutations/asset/helpers.py:286]: bad name. it should correspond to what you do, ie get file_paths

File: kili/mutations/asset/helpers.py [hunk 14]
```diff
@@ -272,4 +281,48 @@ def process_update_properties_in_assets_parameters(properties) -> dict:
...
+            print(files_path)
+        else:
+            files_path.extend([sub_item for sub_item in glob.glob(item)
+                               if os.path.isfile(sub_item)])
+            print(files_path)
+
+    files_path_to_upload = [
+        path for path in files_path if check_file_mime_type(path, input_type, False)]
+    if exclude is not None:
+        files_path_to_upload = sorted(list(
```
Comment (mDuval1) [inline kili/mutations/asset/helpers.py:316]: You do two things here... Filtering And sorting. You must sort separately

File: kili/mutations/asset/helpers.py [hunk 15]
```diff
@@ -272,4 +281,48 @@ def process_update_properties_in_assets_parameters(properties) -> dict:
...
+    if len(files_path_to_upload) == 0:
+        raise ValueError(
+            "No files to upload. "
+            "Check that the paths exist and that the file types are compatible with the project")
+    if len(files_path_to_upload) != len(files_path):
+        unuploaded_files_path = [
+            path for path in files_path if path not in files_path_to_upload]
+        print(
+            f'Files skipped: {unuploaded_files_path}. '
+            'Paths either do not exist or point towards wrong data type for the project')
```
Comment (mDuval1) [inline kili/mutations/asset/helpers.py:327]: or are filtered out

File: test/utils.py [hunk 16]
```diff
@@ -58,3 +59,11 @@ def mocked_count_method(*_):
     Simulates a count query
     """
     return COUNT_SAMPLE_MAX
+
+
+def debug_subprocess_pytest(result):
```
Comment (mDuval1) [inline test/utils.py:64]: used only in test -> move it to test folder

File: kili/cli.py (hunk 9 above)
Comment (theodu) [reply kili/cli.py:58]: indeed, corrected

File: kili/cli.py (hunk 11 above)
Comment (theodu) [reply kili/cli.py:63]: That is an edge case, if people try to mention arguments related to a frame project and that it is not

File: kili/cli.py [hunk 17]
```diff
@@ -12,8 +15,86 @@ def cli():
...
+    """
+    if files is None:
```
[the remark this answers is not in the document]
Comment (PierreLeveau) [reply kili/cli.py:56]: and add typeguard to check that it is  not None at runtime

File: kili/mutations/asset/helpers.py [hunk 18]
```diff
@@ -272,4 +281,48 @@ def process_update_properties_in_assets_parameters(properties) -> dict:
...
+                       input_type: str,
+                       exclude: Optional[Tuple[str, ...]]) -> List[str]:
+    """Get files to upload when giving a list of file or folder paths
+
+    Args:
+        files: a list path that can either be file paths, folder paths or unexisting paths
+        input_type: input type of the project to import data to.
+
+    Returns:
+        a list of existing files to upload, compatible with the project type.
```
Comment (PierreLeveau) [inline kili/mutations/asset/helpers.py:296]: a list of the paths of the files to upload

**Review by PierreLeveau — CHANGES_REQUESTED** [review]

File: kili/cli.py [hunk 19]
```diff
@@ -12,8 +15,86 @@ def cli():
...
+              help="Files to exclude from the given files")
+@click.option('--frames', type=bool, default=False, is_flag=True,
+              help="Only for a frame project, import videos as frames. "
+              "The import time is longer with this option")
+@click.option('--fps', type=int,
+              help="Only for a frame project, import videos with this fps")
+#pylint: disable=too-many-arguments
+def import_assets(api_key: str,
+                  project_id: str,
+                  files: Tuple[str, ...],
```
[the remark this answers is not in the document]
Comment (theodu) [reply kili/cli.py:39]: Click return an argument that can be given several times as a Tuple

File: kili/cli.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (theodu) [reply kili/cli.py:54]: This seems unclear to me. we don't get the idea that if we want to import rich text assets or video from frames, we have to use the SDK

File: test/utils.py (hunk 16 above)
Comment (theodu) [reply test/utils.py:64]: this is already on the test folder

File: kili/cli.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (theodu) [reply kili/cli.py:56]: even better, I can add a required=True in the definition of the argument. thanks

File: kili/cli.py [hunk 20]
```diff
@@ -12,8 +15,86 @@ def cli():
...
+    if input_type == 'FRAME':
+        json_metadata_array = [
+            {'processingParameters': {
+                'shouldKeepNativeFrameRate': fps is None,
+                'framesPlayedPerSecond': fps,
+                'shouldUseNativeVideo': not frames}
+             }
+        ] * len(files_to_upload)
+
+    kili.append_many_to_dataset(
```
[the remark this answers is not in the document]
Comment (theodu) [reply kili/cli.py:89]: I isolated the generation of the json metadata array in a specific function. This cleans a bit the code.
Everything is handled in specific functions.

File: kili/cli.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (PierreLeveau) [reply kili/cli.py:68]: indeed

File: kili/cli.py (hunk 3 above)
Comment (crsegerie) [reply kili/cli.py]: ok

File: kili/cli.py (hunk 1 above)
Comment (crsegerie) [reply kili/cli.py]: @theodu You can maybe add them now to not add to the technical dept

Comment (theodu) [reply kili/cli.py]: @crsegerie I added them, you may be do not look at the last commit

File: kili/cli.py (hunk 11 above)
Comment (mDuval1) [reply kili/cli.py:79]: It concerns the function structure

```
if (edge_case_1): handle case 1
if (edge_case_2): handle case 2
general case
```

File: kili/cli.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (mDuval1) [reply kili/cli.py:96]: ok

File: kili/cli.py (hunk 11 above)
Comment (theodu) [reply kili/cli.py:79]: `if input_type != 'FRAME' and (fps is not None or frames is True)` is an edge case and raise an early error. The general case is defined later
